### ISSUE TITLE
saved model size exceeds 800MB using TF-hub Universal Sentence encoder.

### ISSUE THREAD

**Issue by Aashish-1008** [tensorflow/hub#286]
Hi,

Here is the code.
everything works fine model got exported also as a saved model. 
But the saved model size increased to 870MB.
Is there any way to reduce the model size?

# Implements the Keras model for Text Classification.

`
from tensorflow.python.lib.io import file_io
from tensorflow.keras.layers import Conv1D, Dropout, Dense, Flatten, MaxPooling1D
from tensorflow.keras.models import Model, Sequential
import pandas as pd
import numpy as np
from tensorflow.keras.layers import Conv1D, Dropout, Dense, Flatten, LSTM, MaxPooling1D, Bidirectional
from tensorflow.keras.callbacks import TensorBoard
import tensorflow_hub as hub
from tensorflow.keras.models import load_model
from trainer.utils.fs_resolver import resolve_file_path
Path(estimator.eval_dir()).mkdir(parents=True, exist_ok=True)

import os
import tensorflow as tf
import functools
`

def get_callbacks(config):
    """
    A single callback to save the model h5 & tensor-board training logs during training.
    """

    h5_model_path = os.path.join(resolve_file_path(
        config["ml_cloud_config"]["job_dir"]), config["_id"] + ".h5")
    tb_log_path = os.path.join(resolve_file_path(
        config["ml_cloud_config"]["job_dir"]), 'tensor-board-logs')

    tb_log = TensorBoard(
        log_dir=tb_log_path,
        histogram_freq=0,
        write_graph=True
    )

    checkpoint = tf.keras.callbacks.ModelCheckpoint(h5_model_path,
                                                    save_best_only=True,
                                                    save_weights_only=False,
                                                    verbose=1)
    callbacks = [checkpoint, tb_log]
    return callbacks

def model_fn(training_config):
    """Create a Keras Sequential model with layers.
    Args:
      training_config: (int) model_definition object.
    Returns:
      A Keras model.
    """
    embedding_size = 512
    model = Sequential()

    model.add(Conv1D(32, kernel_size=3, activation='relu', padding='same',
                     input_shape=(embedding_size, 1), name="input"))
    model.add(MaxPooling1D(pool_size=3))
    model.add(Flatten())
    model.add(Dense(256, activation='relu'))
    model.add(Dropout(0.25))
    model.add(Dense(128, activation='relu'))
    model.add(Dropout(0.25))
    model.add(Dense(6, activation='softmax'))

    model.compile(loss='categorical_crossentropy', optimizer='adam',
                  metrics=['accuracy'])
    model.summary()
    return model

def dataset_fn(filename_pattern, config):

    def read_csv_file(filename):
        with file_io.FileIO(filename, 'r') as f:
            df = pd.read_csv(f, header=0, names=['category', 'free_text'])
            return df

    filenames = tf.gfile.Glob(filename_pattern)
    dataframes = [read_csv_file(filename) for filename in filenames]
    df = pd.concat(dataframes)
    text_data = df['free_text'].tolist()[:10]
    label = np.asarray(pd.get_dummies(df["category"]), dtype=np.int8)

    embed_size = embed.get_output_info_dict()['default'].get_shape()[1].value
    embed = get_embed()
    with tf.Session() as session:
        session.run([tf.global_variables_initializer(),
                     tf.tables_initializer()])
        text_data = session.run(embed(text_data))

    text_data = np.array(text_data)
    text_data = np.array([np.reshape(data, (len(data), 1))
                          for data in text_data])

    print(label.shape, text_data.shape)
    return text_data, label[:10]

def get_embed():
  
    print("Loading TF hub module. It might take a while.")
    module_url = "https://tfhub.dev/google/universal-sentence-encoder-large/3"
    return hub.Module(module_url)

def serving_input_receiver_fn():

    input_ph = tf.placeholder(tf.string, shape=[None])
    embed = get_embed()
    embeddings = embed(input_ph)
    embed_tensor = tf.convert_to_tensor(embeddings)
    sentence_embedding = tf.reshape(embed_tensor, [1, 512, 1])
    return tf.estimator.export.ServingInputReceiver({"input_input": sentence_embedding}, {'free_text': input_ph})

def train(config):

    print(config)
    batch_size = 500
    no_epochs = 5
    model = model_fn(config["training_config"])
    model_input_name = model.input_names[0]
    train_inpf = functools.partial(dataset_fn, resolve_file_path(config["featureset"]["test_train"]["train_path"]),
                                   config)
    eval_inpf = functools.partial(dataset_fn, resolve_file_path(config["featureset"]["test_train"]["test_path"]),
                                  config)

    train_spec = tf.estimator.TrainSpec(input_fn=train_inpf, max_steps=2000)
    eval_spec = tf.estimator.EvalSpec(input_fn=eval_inpf)

    estimator = tf.keras.estimator.model_to_estimator(
        keras_model=model,
        model_dir=os.path.join(resolve_file_path(config["ml_cloud_config"]["job_dir"]), 'tensor-board-logs')
    )
    tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)

    model_export_path = resolve_file_path(config["model_export_path"])
    estimator.export_savedmodel(
        model_export_path, serving_input_receiver_fn=serving_input_receiver_fn)
    print("model exported @" + str(model_export_path))

**Comment by eggie5** [comment]
You can try a different embedding w/ a smaller dimensionality for example: `nnlm-en-dim50-with-normalization` vs `nnlm-en-dim128-with-normalization`

**Comment by Aashish-1008** [comment]
> You can try a different embedding w/ a smaller dimensionality for example: `nnlm-en-dim50-with-normalization` vs `nnlm-en-dim128-with-normalization`

But using these embeddings, model performance goes down ie. model is that accurate.
I want to host above model to google cloud ml engine.
But they have this limit of model size should not exceed 250 MB.
What are different options to reduce the size of the model ?

**Comment by Aashish-1008** [comment]
It’s resolved✌️

**Comment by rakeshmothukuru1** [comment]
@Aashish-1008,
Can you please share your resolution so that it will be useful to the community. Thanks in advance.

**Comment by jwsmithers** [comment]
@Aashish-1008, can you mention how you solved this? I'm also struggling with this issue.

**Comment by AjithPanja** [comment]
@rakeshmothukuru1 @jwsmithers  Any update on this issue? I'm also facing the same issue. Thanks in advance :)

**Comment by ZircXes** [comment]
@AjithPanja Have you heard anything - I'm also curious about this.
